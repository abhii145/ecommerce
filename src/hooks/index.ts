import axios from "axios"
import {
  keepPreviousData,
  useQuery,
  UseQueryResult,
} from "@tanstack/react-query"
import { Category, Product, UseProductsParams, ProductResponse } from "../types"

const API_BASE_URL = "https://dummyjson.com/products"

export const useCategories = (): UseQueryResult<Category[]> => {
  return useQuery({
    queryKey: ["categories"],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/categories`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
    placeholderData: keepPreviousData,
  })
}

export const useProducts = ({
  limit,
  skip,
  searchQuery,
  category,
  sortBy,
  order,
}: UseProductsParams): UseQueryResult<ProductResponse> => {
  return useQuery({
    queryKey: ["products", limit, skip, searchQuery, category, sortBy, order],
    queryFn: async () => {
      let url = `${API_BASE_URL}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`

      if (searchQuery) {
        url = `${API_BASE_URL}/search?q=${searchQuery}&limit=${limit}&skip=${skip}`
      } else if (category && category !== "All") {
        url = `${API_BASE_URL}/category/${category}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`
      }

      const response = await axios.get(url)
      return response.data
    },
    placeholderData: keepPreviousData,
    staleTime: 1000 * 60 * 5,
  })
}

export const useSingleProduct = (id: string): UseQueryResult<Product> => {
  return useQuery({
    queryKey: ["singleProducts", id],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/${id}`)
      return response.data
    },
    staleTime: 1000 * 60 * 5,
  })
}

export const useSearchSuggestions = (query: string): UseQueryResult<Product[]> => {
  return useQuery({
    queryKey: ["searchSuggestions", query],
    queryFn: async () => {
      const response = await axios.get(`${API_BASE_URL}/search?q=${query}`)
      return response.data.products.map((product: Product) => ({
        ...product,
        category: product.category || "Uncategorized",
      }))
    },
    enabled: !!query,
    staleTime: 1000 * 60 * 5,
  })
}
