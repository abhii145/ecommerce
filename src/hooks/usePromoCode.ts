import { useDispatch, useSelector } from "react-redux"
import { RootState } from "../store/store"
import {
  setPromoCode,
  clearPromoCode,
  applyPromoCode,
  selectPromoCode,
  clearPromoMessage,
} from "../store/promoCodeSlice"

export const usePromoCode = () => {
  const dispatch = useDispatch()
  const { promoCode, discount, promoMessage } = useSelector(selectPromoCode)
  const { orders } = useSelector((state: RootState) => state.orders)
  const cartItems = useSelector((state: RootState) => state.cart.items)

  const handleApplyPromoCode = () => {
    dispatch(applyPromoCode({ cartItems, orders }))
  }

  const handlePromoCodeChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    resetSelectedCoupon: () => void
  ) => {
    dispatch(setPromoCode(e?.target?.value))
    if (e.target.value === "") {
      dispatch(clearPromoCode())
      resetSelectedCoupon()
    }
  }

  const handleCouponSelect = (couponCode: string) => {
    dispatch(setPromoCode(couponCode))
    dispatch(clearPromoMessage())
  }

  return {
    promoCode,
    discount,
    promoMessage,
    handleApplyPromoCode,
    handlePromoCodeChange,
    handleCouponSelect,
  }
}
