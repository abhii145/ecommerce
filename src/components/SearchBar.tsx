import React, { memo, useState, useEffect, useCallback } from "react"
import { FiSearch, FiX } from "react-icons/fi"
import { Product, SearchBarProps } from "../types"
import { useSearchSuggestions } from "../hooks"
import { useNavigate, Link } from "react-router-dom"
import { debounce } from "lodash"

const MAX_SUGGESTIONS = 10

const SearchBar: React.FC<SearchBarProps> = ({
  searchQuery,
  onSearchChange,
}) => {
  const [query, setQuery] = useState(searchQuery)
  const [debouncedQuery, setDebouncedQuery] = useState(query)
  const { data: suggestions } = useSearchSuggestions(debouncedQuery)
  const navigate = useNavigate()

  const debouncedSetQuery = useCallback(
    (value: string) => {
      debounce(() => {
        setDebouncedQuery(value)
      }, 500)()
    },
    []
  )

  useEffect(() => {
    setQuery(searchQuery)
    debouncedSetQuery(searchQuery)
  }, [searchQuery, debouncedSetQuery])

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    setQuery(value)
    onSearchChange(e)
    debouncedSetQuery(value)
  }

  const handleClearInput = () => {
    setQuery("")
    onSearchChange({ target: { value: "" } } as React.ChangeEvent<HTMLInputElement>)
    setDebouncedQuery("")
  }

  const handleSuggestionClick = (productId: string) => {
    navigate(`/products/${productId}`)
  }

  const groupedSuggestions = suggestions?.reduce((acc, product) => {
    if (!acc[product.category]) {
      acc[product.category] = []
    }
    acc[product.category].push(product)
    return acc
  }, {} as Record<string, Product[]>) || {}

  return (
    <div className="flex-grow mx-4 relative" role="search">
      <FiSearch
        className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400"
        aria-hidden="true"
      />
      <label htmlFor="search-input" className="sr-only">
        Search
      </label>
      <input
        id="search-input"
        type="text"
        value={query}
        onChange={handleInputChange}
        placeholder="Search..."
        className="w-full pl-10 pr-4 py-2 bg-gray-100 rounded-full focus:outline-none text-sm"
        aria-label="Search"
        data-testid="search-input"
      />
      {query && (
        <FiX
          className="absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-400 cursor-pointer"
          aria-hidden="true"
          data-testid="clear-input"
          onClick={handleClearInput}
        />
      )}
      {debouncedQuery && suggestions && suggestions.length > 0 ? (
        <ul
          className="absolute left-0 right-0 max-h-80 overflow-y-auto bg-white border border-gray-200 mt-1 rounded-md shadow-lg z-10 sm:max-w-full sm:w-full"
          data-testid="suggestions-list"
        >
          {Object.entries(groupedSuggestions).map(([category, products]) => (
            <li key={category} className="border-b last:border-b-0">
              <div className="px-4 py-2 bg-gray-100 font-semibold">{category}</div>
              {products.slice(0, MAX_SUGGESTIONS).map((product) => (
                <button
                  key={product.id}
                  onClick={() => handleSuggestionClick(product.id.toString())}
                  className="flex items-center w-full text-left px-4 py-2 hover:bg-gray-100"
                >
                  <img
                    src={product.thumbnail}
                    alt={product.title}
                    className="w-8 h-8 object-cover mr-2"
                  />
                  {product.title}
                </button>
              ))}
            </li>
          ))}
        </ul>
      ) : debouncedQuery && (
        <div className="absolute left-0 right-0 bg-white border border-gray-200 mt-1 rounded-md shadow-lg z-10 px-4 py-2 sm:max-w-full sm:w-full">
          No products found. <Link to="/products" className="text-blue-500">View all products</Link>
        </div>
      )}
    </div>
  )
}

export default memo(SearchBar)
