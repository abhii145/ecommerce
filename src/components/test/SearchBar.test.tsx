import React from "react"
import { render, screen, fireEvent, waitFor } from "@testing-library/react"
import { BrowserRouter as Router } from "react-router-dom"
import SearchBar from "../SearchBar"
import { useSearchSuggestions } from "../../hooks"

jest.mock("../../hooks", () => ({
  useSearchSuggestions: jest.fn(),
}))

const mockUseSearchSuggestions = useSearchSuggestions as jest.Mock

describe("SearchBar", () => {
  const mockOnSearchChange = jest.fn()

  beforeEach(() => {
    mockUseSearchSuggestions.mockReturnValue({ data: [] })
  })

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Router>
        <SearchBar searchQuery="" onSearchChange={mockOnSearchChange} />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders search input", () => {
    render(
      <Router>
        <SearchBar searchQuery="" onSearchChange={mockOnSearchChange} />
      </Router>
    )
    expect(screen.getByTestId("search-input")).toBeInTheDocument()
  })

  it("calls onSearchChange and updates query on input change", () => {
    render(
      <Router>
        <SearchBar searchQuery="" onSearchChange={mockOnSearchChange} />
      </Router>
    )
    const input = screen.getByTestId("search-input")
    fireEvent.change(input, { target: { value: "test" } })
    expect(mockOnSearchChange).toHaveBeenCalled()
    expect(input).toHaveValue("test")
  })

  it("clears input when clear button is clicked", () => {
    render(
      <Router>
        <SearchBar searchQuery="initial" onSearchChange={mockOnSearchChange} />
      </Router>
    )
    const input = screen.getByTestId("search-input")
    const clearButton = screen.getByTestId("clear-input")

    fireEvent.click(clearButton)
    expect(input).toHaveValue("")
    expect(mockOnSearchChange).toHaveBeenCalledWith(expect.objectContaining({ target: { value: "" } }))
  })

  it("shows suggestions when debouncedQuery has results", async () => {
    const suggestions = [
      {
        id: 1,
        title: "Product 1",
        category: "Category 1",
        thumbnail: "img1.jpg",
      },
      {
        id: 2,
        title: "Product 2",
        category: "Category 1",
        thumbnail: "img2.jpg",
      },
    ]
    mockUseSearchSuggestions.mockReturnValue({ data: suggestions })

    render(
      <Router>
        <SearchBar searchQuery="prod" onSearchChange={mockOnSearchChange} />
      </Router>
    )

    const input = screen.getByTestId("search-input")
    fireEvent.change(input, { target: { value: "prod" } })

    await waitFor(() => {
      expect(screen.getByText("Category 1")).toBeInTheDocument()
      expect(screen.getByText("Product 1")).toBeInTheDocument()
      expect(screen.getByText("Product 2")).toBeInTheDocument()
    })
  })

  it("shows 'No products found' when no suggestions are available", async () => {
    render(
      <Router>
        <SearchBar searchQuery="unknown" onSearchChange={mockOnSearchChange} />
      </Router>
    )

    const input = screen.getByTestId("search-input")
    fireEvent.change(input, { target: { value: "unknown" } })

    await waitFor(() => {
      expect(screen.getByText("No products found.")).toBeInTheDocument()
    })
  })

  it("shows 'View all products' link when no suggestions are available", async () => {
    render(
      <Router>
        <SearchBar searchQuery="unknown" onSearchChange={mockOnSearchChange} />
      </Router>
    )

    const input = screen.getByTestId("search-input")
    fireEvent.change(input, { target: { value: "unknown" } })

    await waitFor(() => {
      expect(screen.getByText("No products found.")).toBeInTheDocument()
      expect(screen.getByText("View all products")).toBeInTheDocument()
    })
  })

  it("navigates to product page on suggestion click", async () => {
    const suggestions = [
      {
        id: 1,
        title: "Product 1",
        category: "Category 1",
        thumbnail: "img1.jpg",
      },
    ]
    mockUseSearchSuggestions.mockReturnValue({ data: suggestions })

    render(
      <Router>
        <SearchBar searchQuery="prod" onSearchChange={mockOnSearchChange} />
      </Router>
    )

    const input = screen.getByTestId("search-input")
    fireEvent.change(input, { target: { value: "prod" } })

    await waitFor(() => {
      const suggestionButton = screen.getByText("Product 1")
      fireEvent.click(suggestionButton)
      expect(window.location.pathname).toBe("/products/1")
    })
  })
})
