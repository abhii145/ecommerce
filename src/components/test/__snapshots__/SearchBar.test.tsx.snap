// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-grow mx-4 relative"
    role="search"
  >
    <svg
      aria-hidden="true"
      class="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
    <label
      class="sr-only"
      for="search-input"
    >
      Search
    </label>
    <input
      aria-label="Search"
      class="w-full pl-10 pr-4 py-2 bg-gray-100 rounded-full focus:outline-none text-sm"
      data-testid="search-input"
      id="search-input"
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
